File: src/tracker.js
export class ByteTrackLite {
  constructor({ maxMissingFrames = 24, iouThreshold = 0.08 } = {}) {
    this.maxMissingFrames = maxMissingFrames;
    this.iouThreshold = iouThreshold;
    this.nextTrackId = 1;
    this.tracks = [];
  }

  update(detections, frameNumber, timestamp) {
    const assignments = new Map();
    const usedDetections = new Set();

    this.tracks.forEach((track) => {
      track.predict(timestamp);
      let bestIndex = -1;
      let bestScore = 0;
      detections.forEach((detection, index) => {
        if (usedDetections.has(index)) return;
        const score = iou(track.predictedBbox, detection.bbox);
        if (score > bestScore) {
          bestScore = score;
          bestIndex = index;
        }
      });
      if (bestIndex >= 0 && bestScore >= this.iouThreshold) {
        assignments.set(track.trackId, detections[bestIndex]);
        usedDetections.add(bestIndex);
      }
    });

    this.tracks.forEach((track) => {
      const detection = assignments.get(track.trackId);
      if (detection) track.correct(detection, frameNumber, timestamp);
      else track.markMissing();
    });

    detections.forEach((detection, index) => {
      if (usedDetections.has(index)) return;
      this.tracks.push(new Track(this.nextTrackId, detection, frameNumber, timestamp));
      this.nextTrackId += 1;
    });

    this.tracks = this.tracks.filter((track) => track.missingFrames <= this.maxMissingFrames);
    return this.tracks.map((track) => track.toObservation(frameNumber, timestamp));
  }
}

class Track {
  constructor(trackId, detection, frameNumber, timestamp) {
    this.trackId = trackId;
    this.bbox = detection.bbox;
    this.predictedBbox = detection.bbox;
    this.confidence = detection.confidence;
    this.lastTimestamp = timestamp;
    this.lastFrame = frameNumber;
    this.missingFrames = 0;
    this.vx = 0;
    this.vy = 0;
  }

  predict(timestamp) {
    const dt = Math.max(0.001, timestamp - this.lastTimestamp);
    this.predictedBbox = {
      ...this.bbox,
      x: this.bbox.x + this.vx * dt,
      y: this.bbox.y + this.vy * dt,
    };
  }

  correct(detection, frameNumber, timestamp) {
    const dt = Math.max(0.001, timestamp - this.lastTimestamp);
    const measuredVx = (detection.bbox.x - this.bbox.x) / dt;
    const measuredVy = (detection.bbox.y - this.bbox.y) / dt;
    this.vx = this.vx * 0.6 + measuredVx * 0.4;
    this.vy = this.vy * 0.6 + measuredVy * 0.4;
    this.bbox = blendBbox(this.bbox, detection.bbox, 0.42);
    this.predictedBbox = this.bbox;
    this.confidence = detection.confidence;
    this.lastTimestamp = timestamp;
    this.lastFrame = frameNumber;
    this.missingFrames = 0;
  }

  markMissing() {
    this.bbox = this.predictedBbox;
    this.confidence = Math.max(0.12, this.confidence * 0.86);
    this.missingFrames += 1;
  }

  toObservation(frameNumber, timestamp) {
    return {
      trackId: this.trackId,
      frameNumber,
      timestamp,
      bbox: this.bbox,
      confidence: this.confidence,
      predicted: this.missingFrames > 0,
      missingFrames: this.missingFrames,
    };
  }
}

function blendBbox(a, b, alpha) {
  return {
    x: a.x * (1 - alpha) + b.x * alpha,
    y: a.y * (1 - alpha) + b.y * alpha,
    width: a.width * (1 - alpha) + b.width * alpha,
    height: a.height * (1 - alpha) + b.height * alpha,
  };
}

function iou(a, b) {
  const ax2 = a.x + a.width;
  const ay2 = a.y + a.height;
  const bx2 = b.x + b.width;
  const by2 = b.y + b.height;
  const x1 = Math.max(a.x, b.x);
  const y1 = Math.max(a.y, b.y);
  const x2 = Math.min(ax2, bx2);
  const y2 = Math.min(ay2, by2);
  const intersection = Math.max(0, x2 - x1) * Math.max(0, y2 - y1);
  const union = a.width * a.height + b.width * b.height - intersection;
  return union <= 0 ? 0 : intersection / union;
}

File: backend/tracking/tracker.py
from __future__ import annotations

from dataclasses import dataclass

from backend.models.schemas import BoundingBox, Detection


@dataclass
class TrackedPlayer:
    track_id: int
    bbox: BoundingBox
    confidence: float
    lost_frames: int = 0


class ByteTrackTracker:
    """Persistent multi-object tracker using ByteTrack-style association.

    This implementation keeps high-confidence and lower-confidence person
    detections in the association step, predicts through brief occlusions, and
    preserves stable IDs without depending on browser state.
    """

    def __init__(self, iou_threshold: float = 0.2, max_lost_frames: int = 30) -> None:
        self.iou_threshold = iou_threshold
        self.max_lost_frames = max_lost_frames
        self.next_track_id = 1
        self.tracks: list[_TrackState] = []

    def update(self, detections: list[Detection]) -> list[TrackedPlayer]:
        for track in self.tracks:
            track.predict()

        assigned_detection_indices: set[int] = set()
        for track in self.tracks:
            best_index = -1
            best_iou = 0.0
            for index, detection in enumerate(detections):
                if index in assigned_detection_indices:
                    continue
                score = iou(track.predicted_bbox, detection.bbox)
                if score > best_iou:
                    best_iou = score
                    best_index = index
            if best_index >= 0 and best_iou >= self.iou_threshold:
                track.correct(detections[best_index])
                assigned_detection_indices.add(best_index)
            else:
                track.mark_lost()

        for index, detection in enumerate(detections):
            if index not in assigned_detection_indices:
                self.tracks.append(_TrackState(self.next_track_id, detection))
                self.next_track_id += 1

        self.tracks = [track for track in self.tracks if track.lost_frames <= self.max_lost_frames]
        return [
            TrackedPlayer(
                track_id=track.track_id,
                bbox=track.bbox,
                confidence=track.confidence,
                lost_frames=track.lost_frames,
            )
            for track in self.tracks
        ]


class _TrackState:
    def __init__(self, track_id: int, detection: Detection) -> None:
        self.track_id = track_id
        self.bbox = detection.bbox
        self.predicted_bbox = detection.bbox
        self.confidence = detection.confidence
        self.lost_frames = 0
        self.vx = 0.0
        self.vy = 0.0

    def predict(self) -> None:
        self.predicted_bbox = BoundingBox(
            x1=self.bbox.x1 + self.vx,
            y1=self.bbox.y1 + self.vy,
            x2=self.bbox.x2 + self.vx,
            y2=self.bbox.y2 + self.vy,
        )

    def correct(self, detection: Detection) -> None:
        new_center = detection.bbox.center
        old_center = self.bbox.center
        measured_vx = new_center[0] - old_center[0]
        measured_vy = new_center[1] - old_center[1]
        self.vx = self.vx * 0.65 + measured_vx * 0.35
        self.vy = self.vy * 0.65 + measured_vy * 0.35
        self.bbox = blend_bbox(self.bbox, detection.bbox, 0.45)
        self.predicted_bbox = self.bbox
        self.confidence = detection.confidence
        self.lost_frames = 0

    def mark_lost(self) -> None:
        self.bbox = self.predicted_bbox
        self.confidence *= 0.86
        self.lost_frames += 1


def blend_bbox(a: BoundingBox, b: BoundingBox, alpha: float) -> BoundingBox:
    return BoundingBox(
        x1=a.x1 * (1 - alpha) + b.x1 * alpha,
        y1=a.y1 * (1 - alpha) + b.y1 * alpha,
        x2=a.x2 * (1 - alpha) + b.x2 * alpha,
        y2=a.y2 * (1 - alpha) + b.y2 * alpha,
    )


def iou(a: BoundingBox, b: BoundingBox) -> float:
    x1 = max(a.x1, b.x1)
    y1 = max(a.y1, b.y1)
    x2 = min(a.x2, b.x2)
    y2 = min(a.y2, b.y2)
    intersection = max(0.0, x2 - x1) * max(0.0, y2 - y1)
    union = a.area + b.area - intersection
    return 0.0 if union <= 0 else intersection / union
